File: src/main/java/KNN/KNN.java
package KNN;

import java.io.BufferedReader;
import java.io.File;
import java.io.FileInputStream;
import java.io.FileNotFoundException;
import java.io.InputStreamReader;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.Collections;
import java.util.Comparator;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

public class KNN {	
	double [][] normDataSet;
	
	double [][] returnMat;
	
	double [] classLabelVector;
	
	double [][] returnMatMinMax;
	double [] ranges;
	
	public void print(){
		System.out.print("returnMat:\n[");
		for(int i = 0;i<returnMat.length;i++){
			System.out.print("[");
			
			for(int j=0;j<returnMat[i].length;j++){
				if(j==0){
					System.out.print("");
				}else{
					System.out.print(",");
				}
				System.out.print(returnMat[i][j]);
			}
			
			System.out.println("]");
		}
		System.out.println("]\n");
		
		
		System.out.print("classLabelVector:\n[");
		for(int j=0;j<classLabelVector.length;j++){
			if(j==0){
				System.out.print("");
			}else{
				System.out.print(",");
			}
			System.out.print(classLabelVector[j]);
		}
		System.out.println("]\n");
		
		
		System.out.print("normDataSet:\n[");
		for(int i = 0;i<normDataSet.length;i++){
			System.out.print("[");
			
			for(int j=0;j<normDataSet[i].length;j++){
				if(j==0){
					System.out.print("");
				}else{
					System.out.print(",");
				}
				System.out.print(normDataSet[i][j]);
			}
			
			System.out.println("]");
		}
		System.out.println("]\n");
	}
	
	@Override
	public String toString() {
		return "KNN [returnMat=" + Arrays.toString(returnMat)
				+ ", classLabelVector=" + Arrays.toString(classLabelVector)
				+ "]";
	}

	public void file2matrix(String filename) throws Exception{
		// 打开数据文件，读取每行内容
		BufferedReader textBuffer = new BufferedReader(new InputStreamReader(new FileInputStream(filename)));
		String str = "";
		List<String> list = new ArrayList<String>();
		while ((str = textBuffer.readLine()) != null) {
			list.add(str);
		}
		
		returnMat = new double[list.size()][3];
		classLabelVector = new double[list.size()];
		
		for(int i=0;i<list.size();i++){
			String line = list.get(i);
			String [] dStr = line.trim().split("\t");
			returnMat[i][0] = Double.valueOf(dStr[0].trim()).doubleValue();
			returnMat[i][1] = Double.valueOf(dStr[1].trim()).doubleValue();
			returnMat[i][2] = Double.valueOf(dStr[2].trim()).doubleValue();
			classLabelVector[i] = Double.valueOf(dStr[3].trim()).doubleValue();
		}
		
		
		
/*		while ((str = textBuffer.readLine()) != null) {
			String [] dStr = str.trim().split("\t");
			double [] d = new double[dStr.length];
			for(int i=0;i<dStr.length;i++){
				d[i] = Double.valueOf(dStr[i]);
				System.out.print(d[i]);
				System.out.print("   ");
			}
			System.out.println();
		}*/
	
	}
	
	
	public void autoNorm(){
		
		returnMatMinMax = new double[returnMat[0].length][2];
		for(int i=0;i<returnMatMinMax.length;i++){
			returnMatMinMax[i][0] = 1.7976931348623157E308;
			returnMatMinMax[i][1] = 4.9E-324;
		}
		
		for(int i=0;i<returnMat.length;i++){
			for(int j=0;j<returnMat[i].length;j++){
				if(returnMat[i][j] < returnMatMinMax[j][0]){
					returnMatMinMax[j][0] = returnMat[i][j];
				}
				if(returnMat[i][j] > returnMatMinMax[j][1]){
					returnMatMinMax[j][1] = returnMat[i][j];
				}
			}
		}
		
		ranges = new double[returnMatMinMax.length];
		for(int i=0;i<returnMatMinMax.length;i++){
			ranges[i] = returnMatMinMax[i][1] - returnMatMinMax[i][0];
		}
		
		normDataSet = new double[returnMat.length][3];
		// (旧-小)/(大-小)
		for(int i=0;i<returnMat.length;i++){
			for(int j=0;j<returnMat[i].length;j++){
				normDataSet[i][j] = (returnMat[i][j] - returnMatMinMax[j][0])/ranges[j];
			}
		}
	}
	
	
	public double classify0(double[] intX,double[][] dataSet,double[] labels,int k){
		int dataSetSize = dataSet.length;
		
		double[][] diffMat = new double[dataSet.length][intX.length];
		double[][] sqDiffMat = new double[dataSet.length][intX.length];
		double[] sqDistances = new double[dataSet.length];
		double[] distances = new double[dataSet.length];
		
		List<Map<String,Object>> sortList = new ArrayList<Map<String,Object>>();
		
		for(int i=0;i<dataSet.length;i++){
			// (A1-B1)*(A1-B1)+(A2-B2)*(A2-B2)+(A3-B3)*(A3-B3) 再开平方
			double sum = 0;
			for(int j=0;j<intX.length;j++){
				diffMat[i][j] = intX[j] - dataSet[i][j];
				sqDiffMat[i][j] = Math.pow(diffMat[i][j], 2);
				sum = sum + Math.pow(intX[j] - dataSet[i][j], 2); 
			}
			
			for(int j=0;j<intX.length;j++){
				if(j==0){
					sqDistances[i] = sqDiffMat[i][j];
				}else{
					sqDistances[i] = sqDistances[i] + sqDiffMat[i][j];
				}
			}
			// distances[i] = Math.pow(sqDistances[i],0.5);
			distances[i] = Math.pow(sum,0.5); 
			
			Map<String,Object> o = new HashMap<String,Object>();
			o.put("index", i);
			o.put("value", distances[i]);
			sortList.add(o);
		}
		
		// 
		Collections.sort(sortList, new Comparator<Map<String,Object>>(){

			public int compare(Map<String,Object> arg0, Map<String,Object> arg1) {
			//	return ((Double)arg0.get("value")).compareTo((Double)arg1.get("value"));
				if((Double)arg0.get("value") > (Double)arg1.get("value"))
					return 1;
				else if((Double)arg0.get("value") < (Double)arg1.get("value"))
					return -1;
				else 
					return 0;
			}
			
		});
		
		Map<Double,Integer> classCount = new HashMap<Double,Integer>();
		for(int i=0;i<k;i++){
			HashMap<String,Object> o = (HashMap<String, Object>) sortList.get(i);
			
			double voteIlabel = labels[(Integer)o.get("index")];
			if(classCount.containsKey(voteIlabel)){
				classCount.put(voteIlabel, classCount.get(voteIlabel)+1);
			}else{
				classCount.put(voteIlabel, 1);
			}
		}
		
		// 对classCount排序
		// 通过ArrayList构造函数把map.entrySet()转换成list
        List<Map.Entry<Double,Integer>> maplist = new ArrayList<Map.Entry<Double,Integer>>(classCount.entrySet());
        // 通过比较器实现比较排序
        Collections.sort(maplist, new Comparator<Map.Entry<Double,Integer>>() {
            public int compare(Map.Entry<Double,Integer> mapping1, Map.Entry<Double,Integer> mapping2) {
                return mapping2.getValue().compareTo(mapping1.getValue());
            }
        });
 
        return maplist.get(0).getKey();
	}
	
}

File: src/main/java/KNN/kNN.py
#-*- coding: utf-8 -*-

from numpy import *
import operator

# 读取数据到矩阵
def file2matrix(filename):
    
    # 打开数据文件，读取每行内容
    fr = open(filename)
    arrayOLines = fr.readlines()
    
    # 初始化矩阵
    numberOfLines = len(arrayOLines)
    returnMat = zeros((numberOfLines,3))
    
    # 初始化类标签向量
    classLabelVector = []
    
    # 循环读取每一行数据
    index = 0
    for line in arrayOLines:
        # 去掉回车符
        line = line.strip()
        # 提取4个数据项
        listFromLine = line.split('\t')
        # 将前三项数据存入矩阵
        returnMat[index,:] = listFromLine[0:3]
        # 将第四项数据存入向量
        classLabelVector.append(int(listFromLine[-1]))
        index += 1
    return returnMat,classLabelVector

# 数据归一化
def autoNorm(dataSet):
    # 读取矩阵中数据项的最大和最小值
    minVals = dataSet.min(0)
    maxVals = dataSet.max(0)
    
    # 获得最大和最小值间差值
    ranges = maxVals - minVals
    
    # 初始化输出结果
    normDataSet = zeros(shape(dataSet))
    
    # 获取矩阵的行数
    m = dataSet.shape[0]
    
    # 矩阵运算：实现归一化公式中的 oldValue - min 一步
    normDataSet = dataSet - tile(minVals, (m,1))
    
    # 矩阵除法：实现归一化公式中的除法
    normDataSet = normDataSet/tile(ranges, (m,1))
    
    # 返回归一化后的数据，数据范围及最小值矩阵
    return normDataSet, ranges, minVals

# kNN算法实现    
def classify0(inX, dataSet, labels, k):
    # 获取样本数据数量
    dataSetSize = dataSet.shape[0]
    
    # 矩阵运算，计算测试数据与每个样本数据对应数据项的差值
    diffMat = tile(inX, (dataSetSize,1)) - dataSet
    
    # sqDistances 上一步骤结果平方和
    sqDiffMat = diffMat**2
    sqDistances = sqDiffMat.sum(axis=1)
    
    # 取平方根，得到距离向量
    distances = sqDistances**0.5
    
    # 按照距离从低到高排序
    sortedDistIndicies = distances.argsort()     
    classCount={}          
    
    # 依次取出最近的样本数据
    for i in range(k):
        # 记录该样本数据所属的类别
        voteIlabel = labels[sortedDistIndicies[i]]
        classCount[voteIlabel] = classCount.get(voteIlabel,0) + 1
    
    # 对类别出现的频次进行排序，从高到低
    sortedClassCount = sorted(classCount.iteritems(), key=operator.itemgetter(1), reverse=True)
    
    # 返回出现频次最高的类别
    return sortedClassCount[0][0]

# 算法测试    
def datingClassTest():
    # 设定测试数据的比例
    hoRatio = 0.10
    
    # 读取数据
    datingDataMat,datingLabels = file2matrix('datingTestSet2.txt')
    
    # 归一化数据
    normMat, ranges, minVals = autoNorm(datingDataMat)
    
    # 数据总行数
    m = normMat.shape[0]
    
    # 测试数据行数
    numTestVecs = int(m*hoRatio)
    
    # 初始化错误率
    errorCount = 0.0
    
    # 循环读取每行测试数据
    for i in range(numTestVecs):
    
        # 对该测试人员进行分类
        classifierResult = classify0(normMat[i,:],normMat[numTestVecs:m,:],datingLabels[numTestVecs:m],3)
        
        # 打印KNN算法分类结果和真实的分类
        print "the classifier came back with: %d, the real answer is: %d" % (classifierResult, datingLabels[i])
        
        # 判断KNN算法结果是否准确
        if (classifierResult != datingLabels[i]): errorCount += 1.0
    
    # 打印错误率
    print "the total error rate is: %f" % (errorCount/float(numTestVecs))

# 执行算法测试
datingClassTest()